# Optimization of Oil Exploration: Predictive Modeling and Risk Analysis for Selecting Profitable Wells

## 1.- General Information

### 1.1.- Initialization

In [71]:
import pandas as pd
import numpy as np
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, accuracy_score, classification_report, confusion_matrix
from sklearn.metrics import roc_auc_score
from sklearn.metrics import f1_score
from sklearn.dummy import DummyClassifier
from sklearn.utils import shuffle, resample
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score


In [72]:
region_1 = pd.read_csv("../geo_data_0.csv")
region_2 = pd.read_csv("../geo_data_1.csv")
region_3 = pd.read_csv("../geo_data_2.csv")

### 1.2.- Preprocessing data

In [73]:
region_1 = pd.read_csv("../geo_data_0.csv")
region_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [74]:
id_duplicates = region_1["id"].duplicated().sum()
print(f"Duplicated ID in region 1: {id_duplicates}")


# Verify if rows are totally duplicated
rows_duplicates = region_1.duplicated().sum()
print(f"Rows duplicated in region 1: {rows_duplicates}")


id_duplicates = region_2["id"].duplicated().sum()
print(f"Duplicated ID in region 2: {id_duplicates}")


# Verify if rows are totally duplicated
rows_duplicates = region_2.duplicated().sum()
print(f"Rows duplicated in region 2: {rows_duplicates}")

id_duplicates = region_3["id"].duplicated().sum()
print(f"Duplicated ID in region 3: {id_duplicates}")


# Verify if rows are totally duplicated
rows_duplicates = region_3.duplicated().sum()
print(f"Rows duplicated in region 3: {rows_duplicates}")

Duplicated ID in region 1: 10
Rows duplicated in region 1: 0
Duplicated ID in region 2: 4
Rows duplicated in region 2: 0
Duplicated ID in region 3: 4
Rows duplicated in region 3: 0


In [75]:
# Eliminar duplicados manteniendo la primera aparición
region_1 = region_1.drop_duplicates(subset="id", keep="first")
# Verificar si aún hay duplicados
print(region_1["id"].duplicated().sum())  # Debería imprimir 0

# Eliminar duplicados manteniendo la primera aparición
region_2 = region_2.drop_duplicates(subset="id", keep="first")
# Verificar si aún hay duplicados
print(region_2["id"].duplicated().sum())  

# Eliminar duplicados manteniendo la primera aparición
region_3 = region_3.drop_duplicates(subset="id", keep="first")
# Verificar si aún hay duplicados
print(region_3["id"].duplicated().sum())  

0
0
0


## 2.- Model

### Region 1

In [76]:
target = region_1['product']
features = region_1.drop(['product','id'], axis=1)

features_train, features_validation, target_train, target_validation = train_test_split(
    features, target, test_size=0.25, random_state=12345
)

# Size verification
print(f"Train data size: {len(features_train)}")
print(f"Validation data size: {len(features_validation)}")


Train data size: 74992
Validation data size: 24998


In [77]:
model = LinearRegression()

# Train the model using the training data
model.fit(features_train, target_train)

# Make predictions on the validation set
predictions_region_1 = model.predict(features_validation)

results_df = pd.DataFrame({
    "Actual": target_validation.values,
    "Predicted": predictions_region_1
})

# Calculate mean predicted reserves volume
mean_predicted_volume = predictions_region_1.mean()

print('Region 1 results:')
print(f"Mean Predicted Reserves Volume: {mean_predicted_volume:.2f} (thousands of barrels)")

# Calculate evaluation metrics
mse = mean_squared_error(target_validation, predictions_region_1)  # Mean Squared Error
rmse = mse ** 0.5  # Root Mean Squared Error
r2 = r2_score(target_validation, predictions_region_1)  # R² Score

# Print results
print(f"Mean Squared Error (MSE): {mse:.2f}")
print(f"Root Mean Squared Error (RMSE): {rmse:.2f}")
print(f"R² Score: {r2:.2f}")



Region 1 results:
Mean Predicted Reserves Volume: 92.79 (thousands of barrels)
Mean Squared Error (MSE): 1432.89
Root Mean Squared Error (RMSE): 37.85
R² Score: 0.27


#### Model Performance Analysis (Region 1)

- The model's performance is **low**, with an **R² score of 0.27**, meaning the model only explains **27% of the variance** in the target variable (`product`).
- The **high RMSE (37.85)** indicates significant errors in predictions, suggesting that the model struggles to capture the relationship between features and reserves.
- The **Mean Squared Error (MSE) of 1432.89** reinforces the presence of large prediction errors.
- The **Mean Predicted Reserves Volume is 92.79 thousand barrels**, but the high RMSE suggests that individual predictions may vary widely.
- This performance suggests that **linear regression may not be the best fit for this region** and that further data preprocessing, feature selection, or alternative modeling approaches might be needed.

### Region 2

In [78]:
target = region_2['product']
features = region_2.drop(['product','id'], axis=1)

features_train, features_validation, target_train, target_validation = train_test_split(
    features, target, test_size=0.25, random_state=12345
)

print(f"Train data size: {len(features_train)}")
print(f"Validation data size: {len(features_validation)}")

Train data size: 74997
Validation data size: 24999


In [79]:
model = LinearRegression()

# Train the model using the training data
model.fit(features_train, target_train)

# Make predictions on the validation set
predictions_region_2 = model.predict(features_validation)

results_df = pd.DataFrame({
    "Actual": target_validation.values,
    "Predicted": predictions_region_2
})

# Calculate mean predicted reserves volume
mean_predicted_volume = predictions_region_2.mean()

print('Region 2 results:')

print(f"Mean Predicted Reserves Volume: {mean_predicted_volume:.2f} (thousands of barrels)")

# Calculate evaluation metrics
mse = mean_squared_error(target_validation, predictions_region_2)  # Mean Squared Error
rmse = mse ** 0.5  # Root Mean Squared Error
r2 = r2_score(target_validation, predictions_region_2)  # R² Score

# Print results
print(f"Mean Squared Error (MSE): {mse:.2f}")
print(f"Root Mean Squared Error (RMSE): {rmse:.2f}")
print(f"R² Score: {r2:.2f}")



Region 2 results:
Mean Predicted Reserves Volume: 69.18 (thousands of barrels)
Mean Squared Error (MSE): 0.80
Root Mean Squared Error (RMSE): 0.89
R² Score: 1.00


#### Model Performance Analysis (Region 2)

- The model's performance is **excellent**, with an **R² score of 1.00**, meaning the model explains **100% of the variance** in the target variable (`product`).
- The **low RMSE (0.89)** indicates that the model's predictions are highly accurate, with minimal errors.
- The **Mean Squared Error (MSE) of 0.80** further confirms the model's precision.
- The **Mean Predicted Reserves Volume is 69.18 thousand barrels**, suggesting that the model predicts a consistent volume for new wells.
- Given this perfect performance, the model may be **overfitting**, meaning it might be memorizing the training data instead of generalizing well for new data. Further validation on unseen data is recommended.


### Region 3

In [80]:
target = region_3['product']
features = region_3.drop(['product','id'], axis=1)

features_train, features_validation, target_train, target_validation = train_test_split(
    features, target, test_size=0.25, random_state=12345
)

# Size verification
print(f"Train data size: {len(features_train)}")
print(f"Validation data size: {len(features_validation)}")
#print(f"Test data size: {len(features_test)}")

Train data size: 74997
Validation data size: 24999


In [81]:
model = LinearRegression()

# Train the model using the training data
model.fit(features_train, target_train)

# Make predictions on the validation set
predictions_region_3 = model.predict(features_validation)

results_df = pd.DataFrame({
    "Actual": target_validation.values,
    "Predicted": predictions_region_3
})

# Calculate mean predicted reserves volume
mean_predicted_volume = predictions_region_3.mean()

print('Region 3 results:')
print(f"Mean Predicted Reserves Volume: {mean_predicted_volume:.2f} (thousands of barrels)")

# Calculate evaluation metrics
mse = mean_squared_error(target_validation, predictions_region_3)  # Mean Squared Error
rmse = mse ** 0.5  # Root Mean Squared Error
r2 = r2_score(target_validation, predictions_region_3)  # R² Score

# Print results
print(f"Mean Squared Error (MSE): {mse:.2f}")
print(f"Root Mean Squared Error (RMSE): {rmse:.2f}")
print(f"R² Score: {r2:.2f}")



Region 3 results:
Mean Predicted Reserves Volume: 94.87 (thousands of barrels)
Mean Squared Error (MSE): 1606.07
Root Mean Squared Error (RMSE): 40.08
R² Score: 0.20


#### Model Performance Analysis (Region 3)

- The model's performance is **low**, with an **R² score of 0.20**, meaning the model only explains **20% of the variance** in the target variable (`product`).
- The **high RMSE (40.08)** suggests significant errors in predictions, indicating poor accuracy.
- The **Mean Squared Error (MSE) of 1606.07** confirms that the model struggles to fit the data properly.
- The **Mean Predicted Reserves Volume is 94.87 thousand barrels**, but the high RMSE implies that individual predictions may vary significantly.
- This weak performance suggests that **linear regression may not be the best model for this region**, or that additional feature engineering and data preprocessing (e.g., removing outliers, normalizing features) might improve results.

## 3.- Benefits

In [82]:
# Define key business parameters
BUDGET = 100_000_000  # Total budget in dollars
WELLS_SELECTED = 200  # Wells to be selected
REVENUE_PER_UNIT = 4_500  # Revenue per unit of product (1000 barrels)
BREAK_EVEN_UNITS = 111.1  # Minimum units required per well to avoid losses

# Calculate the minimum revenue required per well
MIN_REVENUE_PER_WELL = BUDGET / WELLS_SELECTED  # $500,000 per well
MIN_UNITS_REQUIRED = MIN_REVENUE_PER_WELL / REVENUE_PER_UNIT  # 111.1 reserve units

# Display the profitability threshold
print(f"Each well must have at least {MIN_UNITS_REQUIRED:.1f} thousand barrels to avoid losses.")


Each well must have at least 111.1 thousand barrels to avoid losses.


### 3.1.- Top 200 wells

In [83]:
def calculate_profit(predicted_volumes):

    total_units = np.sum(predicted_volumes)  # Total sum of reserve units
    revenue = total_units * REVENUE_PER_UNIT  # Revenue calculation
    profit = revenue - BUDGET  # Net profit after investment
    
    return profit


In [84]:
# Top 200 wells
top_200_region_1 = np.sort(predictions_region_1)[-WELLS_SELECTED:]
top_200_region_2 = np.sort(predictions_region_2)[-WELLS_SELECTED:]
top_200_region_3 = np.sort(predictions_region_3)[-WELLS_SELECTED:]

# Calculate profit for each region
profit_region_1 = calculate_profit(top_200_region_1)
profit_region_2 = calculate_profit(top_200_region_2)
profit_region_3 = calculate_profit(top_200_region_3)

# Show results
print(f"Expected profit Region 1: ${profit_region_1:,.2f}")
print(f"Expected profit Region 2: ${profit_region_2:,.2f}")
print(f"Expected profit Region 3: ${profit_region_3:,.2f}")


Expected profit Region 1: $39,816,309.11
Expected profit Region 2: $24,860,607.85
Expected profit Region 3: $33,629,149.08


In [85]:
profits = {
    "Region 1": profit_region_1,
    "Region 2": profit_region_2,
    "Region 3": profit_region_3
}

# Find the best region
best_region = max(profits, key=profits.get)
print(f"The best region to drill is: {best_region} with an expected profit of ${profits[best_region]:,.2f}.")

The best region to drill is: Region 1 with an expected profit of $39,816,309.11.


### 3.2.-Risk and profit

In [86]:
def bootstrap_profit(predictions, n_samples=1000):
    """
    Performs bootstrapping to estimate profit distribution.
    
    Parameters:
    - predictions: array of predicted reserve volumes
    - n_samples: number of bootstrap iterations (default: 1000)
    
    Returns:
    - Mean expected profit
    - 95% confidence interval
    - Probability of loss (as a percentage)
    """
    profits = []
    
    for sample in range(n_samples):
        sampled_wells = np.random.choice(predictions, size=WELLS_SELECTED, replace=True)  # Random sample
        profit = calculate_profit(sampled_wells)  # Calculate profit for this sample
        profits.append(profit)
    
    profits = np.array(profits)
    
    # Calculate statistics
    mean_profit = np.mean(profits)  # Average expected profit
    lower_bound, upper_bound = np.percentile(profits, [2.5, 97.5])  # 95% confidence interval
    loss_risk = (profits < 0).mean() * 100  # Probability of negative profit (as a percentage)
    
    return mean_profit, (lower_bound, upper_bound), loss_risk

# Apply bootstrapping to each region
mean_profit_1, conf_interval_1, loss_risk_1 = bootstrap_profit(top_200_region_1)
mean_profit_2, conf_interval_2, loss_risk_2 = bootstrap_profit(top_200_region_2)
mean_profit_3, conf_interval_3, loss_risk_3 = bootstrap_profit(top_200_region_3)

# Display results
print(f"Region 1: Expected Profit = ${mean_profit_1:,.2f}, 95% CI = {conf_interval_1}, Risk of Loss = {loss_risk_1:.2f}%")
print(f"Region 2: Expected Profit = ${mean_profit_2:,.2f}, 95% CI = {conf_interval_2}, Risk of Loss = {loss_risk_2:.2f}%")
print(f"Region 3: Expected Profit = ${mean_profit_3:,.2f}, 95% CI = {conf_interval_3}, Risk of Loss = {loss_risk_3:.2f}%")


Region 1: Expected Profit = $39,817,209.30, 95% CI = (39075154.68985274, 40614263.0477617), Risk of Loss = 0.00%
Region 2: Expected Profit = $24,860,898.46, 95% CI = (24817237.291371483, 24900974.955437575), Risk of Loss = 0.00%
Region 3: Expected Profit = $33,612,335.45, 95% CI = (32873007.594096687, 34334083.21033316), Risk of Loss = 0.00%


In [87]:
# Store results in a dictionary
results = {
    "Region 1": {"Profit": mean_profit_1, "CI": conf_interval_1, "Risk": loss_risk_1},
    "Region 2": {"Profit": mean_profit_2, "CI": conf_interval_2, "Risk": loss_risk_2},
    "Region 3": {"Profit": mean_profit_3, "CI": conf_interval_3, "Risk": loss_risk_3}
}

# Print final recommendation
print(f"The best region for oil extraction is {best_region}, with an expected profit of ${best_profit:,.2f}.")
print(f"Risk of loss: {best_risk:.2f}% (below the 2.5% threshold).")
print(f"95% confidence interval: {best_ci}")



The best region for oil extraction is Region 1, with an expected profit of $39,789,460.50.
Risk of loss: 0.00% (below the 2.5% threshold).
95% confidence interval: (39055151.21062289, 40575458.53052304)


## Conclusions





### 📌 Final Conclusion: Best Region for Oil Extraction

#### **📊 Profit and Risk Analysis**
- The best region for oil extraction is **Region 1**, with an **expected profit of $39,789,460.50**.
- The **risk of loss is 0.00%**, which is well below the **2.5% threshold**, ensuring a highly stable investment.
- The **95% confidence interval** for expected profit is **($39,055,151.21, $40,575,458.53)**, indicating a narrow and reliable range.

#### **✅ Decision Justification**
- Region 1 provides the **highest profit potential** with virtually no risk of loss.
- The confidence interval suggests **low variability**, making this a **secure investment**.
- Given these results, **Region 1 is strongly recommended** for developing the 200 new oil wells.

🚀 **Next Steps:**
- Move forward with planning and resource allocation for drilling in **Region 1**.
- Continue monitoring oil price fluctuations and potential geological factors to ensure long-term profitability.
